### PR TITLE
Hloc fix for image_ids (camera per image)

### PR DESCRIPTION
**Pull request by dberga**
Hloc gets a single image_ids [here](https://github.com/cvg/Hierarchical-Localization/blob/master/hloc/reconstruction.py#L129) so that it is unable to get the matching between images. Thus it gets an error on not finding images "frame_XXX not found". This little fix in `hloc_utils.py` is solving this issue https://github.com/nerfstudio-project/nerfstudio/issues/3059

Found this error on using `--no_same_dimensions` with `disk` features and `disk+lightglue` matching

### CODE REVIEW

**Comment by dberga** [comment]
This was a dependency problem

error: Failed to download: protobuf==3.12.1
  Caused by: HTTP status server error (503 Service Unavailable) for url (https://files.pythonhosted.org/packages/e0/6f/076967445a5c45d01b75b1a873e8e8c99a26fc54c279b7536eb3be726196/protobuf-3.12.1-cp38-cp38-manylinux1_x86_64.whl.metadata)

Not related to the commit. 

Can you re-run the Build Core test?

**Comment by jb-ye** [comment]
Let me recap my understanding of the issue. Are you capturing horizontal pictures and portrait pictures with the same resolution? In the original intention, `--no_same_dimensions` was intended for rotating portrait mode images by 90 degrees such that all images are of the same resolution for colmap/hloc. 

To my knowledge, the current preprocessing in nerfstudio doesn't support distinctive image resolutions.

**Comment by dberga** [comment]
> Let me recap my understanding of the issue. Are you capturing horizontal pictures and portrait pictures with the same resolution? In the original intention, `--no_same_dimensions` was intended for rotating portrait mode images by 90 degrees such that all images are of the same resolution for colmap/hloc.
> 
> To my knowledge, the current preprocessing in nerfstudio doesn't support distinctive image resolutions.

If the camera parameters are defined for every image (frame) the nerfstudio parser can read each camera params independently as in (https://github.com/nerfstudio-project/nerfstudio/blob/main/nerfstudio/data/dataparsers/nerfstudio_dataparser.py#L83) defined as a bool `distort_fixed` that detects that. The `transforms.json` then looks like https://github.com/InternLandMark/LandMark?tab=readme-ov-file#prepare-dataset.
In the `colmap_to_json.py` I am working on making the https://github.com/nerfstudio-project/nerfstudio/blob/main/nerfstudio/process_data/colmap_utils.py#L390 export multifocal in case there are distinct cameras.

In the `hloc_utils.py` (https://github.com/nerfstudio-project/nerfstudio/pull/3081/commits/2a9b86f504e34c7174e699bb7bc7d2004cc4115d#diff-f6c061630b26a6080b7a03cb99f1dfcd3831b10ceaed217afd958a4a0bbcff26R150) by default the `image_list` is not defined in the code so that it gets error on looking for the images. The commit also fixes that mainly.

A possible solution to avoid all this is to keep all the images with the same resolution by padding the borders. 

How can the Core fail be fixed? It is related to the dependency, not related to the code :(

**Comment by dberga** [comment]
Can the Core build be re-run?

Now it's pip who couldn't fetch another wheel: botocore==1.34.85 ...

**Comment by jb-ye** [comment]
okay, I think my main concern here is if the user actually have images from the same camera model, would it makes sense to use `camera_mode=pycolmap.CameraMode.SINGLE`? Could you add a parameter to `run_hloc()` as well as `class ColmapConverterToNerfstudioDataset` to allow `pycolmap.CameraMode.PER_IMAGE` be an option rather than a default choice?

**Comment by jb-ye** [comment]
> Can the Core build be re-run?

CI is not super stable, I guess. Don't worry about it.

**Comment by dberga** [comment]
@jb-ye just pushed a commit with the required argument. also added the necessary code to export the multifocal case (similarly as in the nerfstudio dataparser distortion_fixed bool) in the colmap_to_json. Check this: https://github.com/nerfstudio-project/nerfstudio/pull/3081/commits/1f73743bfe3dd2602901d752d14d7104f10973ed

image

File: nerfstudio/process_data/colmap_converter_to_nerfstudio_dataset.py [hunk 1]
```diff
@@ -101,6 +101,8 @@ class ColmapConverterToNerfstudioDataset(BaseConverterToNerfstudioDataset):
     """If --use-sfm-depth and this flag is True, also export debug images showing Sf overlaid upon input images."""
     same_dimensions: bool = True
     """Whether to assume all images are same dimensions and so to use fast downscaling with no autorotation."""
+    same_camera: bool = True
```
Comment (jb-ye) [inline nerfstudio/process_data/colmap_converter_to_nerfstudio_dataset.py:104]: maybe use the variable name to `use_single_camera_mode`

File: nerfstudio/process_data/colmap_utils.py [hunk 2]
```diff
@@ -418,6 +418,14 @@ def colmap_to_json(
     # im_id_to_image = recon.images
     cam_id_to_camera = read_cameras_binary(recon_dir / "cameras.bin")
     im_id_to_image = read_images_binary(recon_dir / "images.bin")
+    multifocal = False  # one camera for all frames (distort_fixed=True)
+    if set(cam_id_to_camera.keys()) != {1}:
+        print("Only single camera shared for all images is supported.")
```
Comment (jb-ye) [inline nerfstudio/process_data/colmap_utils.py:423]: Given the change, I think now we are parsing multiple cameras from colmap. So just give a warning message indicating that more than one camera is found in the colmap project should suffice. We can set `out={}` here.

File: nerfstudio/process_data/colmap_utils.py [hunk 3]
```diff
@@ -418,6 +418,14 @@ def colmap_to_json(
     # im_id_to_image = recon.images
     cam_id_to_camera = read_cameras_binary(recon_dir / "cameras.bin")
     im_id_to_image = read_images_binary(recon_dir / "images.bin")
+    multifocal = False  # one camera for all frames (distort_fixed=True)
```
Comment (jb-ye) [inline nerfstudio/process_data/colmap_utils.py:421]: Maybe use `use_single_camera_mode = True`

**Comment by jb-ye** [comment]
@dberga LGTM, I just have some minor comments about the change.

**Comment by dberga** [comment]
> @dberga LGTM, I just have some minor comments about the change.

Right, will do the corrections asap during the next week :)

**Comment by dberga** [comment]
> @dberga LGTM, I just have some minor comments about the change.

Just made all the changes in https://github.com/nerfstudio-project/nerfstudio/pull/3081/commits/b7573d5a5ac02c7a3c71342f36da58ee661f6616

File: nerfstudio/process_data/colmap_utils.py [hunk 4]
```diff
@@ -455,10 +463,12 @@ def colmap_to_json(
         if image_id_to_depth_path is not None:
             depth_path = image_id_to_depth_path[im_id]
             frame["depth_file_path"] = str(depth_path.relative_to(depth_path.parent.parent))
+
+        if use_single_camera_mode is True:  # add the camera for this frame
```
Comment (jb-ye) [inline nerfstudio/process_data/colmap_utils.py:467]: Should this be False?

**Review by jb-ye — CHANGES_REQUESTED** [review]

File: nerfstudio/process_data/colmap_utils.py (hunk 4 above)
Comment (dberga) [reply nerfstudio/process_data/colmap_utils.py]: Just fixed this in https://github.com/nerfstudio-project/nerfstudio/pull/3081/commits/999b7617ce833ead75ab2c553c1bc2231c0e1215

File: nerfstudio/process_data/colmap_utils.py [hunk 5]
```diff
@@ -455,10 +463,12 @@ def colmap_to_json(
...
+            frame.update(parse_colmap_camera_params(cam_id_to_camera[im_id]))
```
[the remark this answers is not in the document]
Comment (dberga) [reply nerfstudio/process_data/colmap_utils.py:468]: Just fixed https://github.com/nerfstudio-project/nerfstudio/pull/3081/commits/999b7617ce833ead75ab2c553c1bc2231c0e1215

[the remark this answers is not in the document]
Comment (jb-ye) [reply nerfstudio/process_data/colmap_utils.py:468]: The im_id should instead be camera id, right? I don’t see the fix in the latest commit. Did I miss anything?

[the remark this answers is not in the document]
Comment (jb-ye) [reply nerfstudio/process_data/colmap_utils.py:468]: The remaining change is simple, let me take care of this forward to merging this PR. No worry @dberga

**Comment by jb-ye** [comment]
@dberga I couldn't push my changes to your feature branch, could you either change the permission to allow upstream authors (or just me) or make the suggested change accordingly. 
```
-        if use_single_camera_mode is False:  # add the camera for this frame
-            frame.update(parse_colmap_camera_params(cam_id_to_camera[im_id]))
+        if not use_single_camera_mode:  # add the camera parameters for this frame
+            frame.update(parse_colmap_camera_params(cam_id_to_camera[im_data.camera_id]))
```

**Comment by dberga** [comment]
> @dberga I couldn't push my changes to your feature branch, could you either change the permission to allow upstream authors (or just me) or make the suggested change accordingly.
> 
> ```
> -        if use_single_camera_mode is False:  # add the camera for this frame
> -            frame.update(parse_colmap_camera_params(cam_id_to_camera[im_id]))
> +        if not use_single_camera_mode:  # add the camera parameters for this frame
> +            frame.update(parse_colmap_camera_params(cam_id_to_camera[im_data.camera_id]))
> ```

Done !

**Review by jb-ye — APPROVED** [review]
I manually tested this PR with a local colmap project that have multiple cameras, it works as expected.

**Comment by ichsan2895** [comment]
Hello, I success for using hloc with this command
```
ns-process-data images --data IMAGES \
    --output-dir IMAGES_MULTICAM \
    --sfm-tool hloc --feature-type disk --matcher-type disk+lightglue \
    --no-same-dimensions --no-use_single_camera_mode
```
it detected all images (just 8 images)
image
image
image


But... When I run `splatfacto`,
the `ns-viewer` only detected 7 poses. There are no eval images too (AFAIK, 8th images should be eval)
image

Please gives me insight @jb-ye @dberga

**Comment by dberga** [comment]
> Hello, I success for using hloc with this command
> 
> ```
> ns-process-data images --data IMAGES \
>     --output-dir IMAGES_MULTICAM \
>     --sfm-tool hloc --feature-type disk --matcher-type disk+lightglue \
>     --no-same-dimensions --no-use_single_camera_mode
> ```
> 
> it detected all images (just 8 images) image image image
> 
> But... When I run `splatfacto`, the `ns-viewer` only detected 7 poses. There are no eval images too (AFAIK, 8th images should be eval) image
> 
> Please gives me insight @jb-ye @dberga

It will depend on the amount of matches found from the features detected in the colmap/hloc method. 
Considering there are 8 images, the disk+lightglue could find 7 cameras but maybe one of the images didn't watch with any other, thus being discarded. As being discarded the viewer can only show the available cameras. 

Can you share the transforms.json? Is there any frame missing?

To solve the lack on matching please try to add some parameters on the ns-process data methods (e.g. other feature extractors and matchers, enable exhaustive feature search and perspective mode, etc.).

**Comment by ichsan2895** [comment]
Hello @dberga , I want to share my `transforms.json`.
Additional info:
```
ns-process-data images --data IMAGES \
    --output-dir IMAGES_MULTICAM \
    --sfm-tool hloc --feature-type disk --matcher-type disk+lightglue \
    --skip-images-processing --no-same-dimensions --no-use_single_camera_mode
```
detected 8 of 8  images (100%).
[transforms.json](https://github.com/nerfstudio-project/nerfstudio/files/15113185/transforms.json)

Second attempt:
```
ns-process-data images --data IMAGES \
    --output-dir IMAGES_MULTICAM \
    --sfm-tool hloc --feature-type disk --matcher-type disk+lightglue \
--no-same-dimensions --no-use_single_camera_mode
```
detected 7 of 8  images. (87,5%)
[transforms (1).json](https://github.com/nerfstudio-project/nerfstudio/files/15113233/transforms.1.json)

colmap 3.9.1 with CUDA, built from source
installed latest commits of `nerfstudio` and `gsplat`

**Comment by jb-ye** [comment]
> the ns-viewer only detected 7 poses. There are no eval images too (AFAIK, 8th images should be eval)
Could you share your dataset?

**Comment by ichsan2895** [comment]
> > the ns-viewer only detected 7 poses. There are no eval images too (AFAIK, 8th images should be eval)
> 
> Could you share your dataset? @ichsan2895

Here is it
https://drive.usercontent.google.com/download?id=1ChzTC_Vokc5ms0zIW23VH76maHUSq44p&authuser=0

**Comment by dberga** [comment]
> Hello @dberga , I want to share my `transforms.json`. Additional info: **First Attempt**
> 
> ```
> ns-process-data images --data IMAGES \
>     --output-dir IMAGES_MULTICAM \
>     --sfm-tool hloc --feature-type disk --matcher-type disk+lightglue \
>     --skip-images-processing --no-same-dimensions --no-use_single_camera_mode
> 
> cp -r IMAGES/* IMAGES_MULTICAM/images
> ```
> 
> detected 8 of 8 images (100%). [transforms.json](https://github.com/nerfstudio-project/nerfstudio/files/15113185/transforms.json)
> 
> **Second attempt:**
> 
> ```
> ns-process-data images --data IMAGES \
>     --output-dir IMAGES_ORI-MULTICAM \
>     --sfm-tool hloc --feature-type disk --matcher-type disk+lightglue \
>     --no-same-dimensions --no-use_single_camera_mode
> ```
> 
> detected 7 of 8 images. (87,5%) [transforms (1).json](https://github.com/nerfstudio-project/nerfstudio/files/15113233/transforms.1.json)
> 
> colmap 3.9.1 with CUDA, built from source hloc-1.4 installed latest commits of `nerfstudio` and `gsplat`

I see in the first case that you add `--skip-images-processing`, in the first case. By default without downscaling (skipping images processing) won't create the "images_X" folders with "frame_X" files. Without specifying this parameter it downscales by x2, x4, x8. In principle as the preprocessing (hloc+disk features+lightglue) works best with higher resolutions. 

Can you check (with COLMAP gui) the database that all cameras are found? Did the hloc found matches for all cases? In the second transforms case the frame_0001.jpg is missing.

**Comment by ichsan2895** [comment]
Yes, I was intended to use `--skip-image-proprecessing` and manually copy 8 images to `IMAGE_MULTICAM/images`. I use `splatfacto` with `--downscale-factor 1`, so it does not need `images_2`, `images_4`, and `images_8`.

**Comment by ichsan2895** [comment]
Hey, Sorry for my careless. In first attempt that I said before:
```
ns-process-data images --data IMAGES \
    --output-dir IMAGES_MULTICAM \
    --sfm-tool hloc --feature-type disk --matcher-type disk+lightglue \
    --skip-images-processing --no-same-dimensions --no-use_single_camera_mode

cp -r IMAGES/* IMAGES_MULTICAM/images
```
I found 8th images in `ns-view` with `viser`, but it has wrong orientation.
image

Take a look in portrait image, but in `viser`, it was rotated automatically.
image

The `transforms.json` also does not translate it to potrait.
image

**Comment by dberga** [comment]
> Yes @dberga , I was intended to use `--skip-image-proprecessing` and manually copy 8 images to `IMAGE_MULTICAM/images`. I use `splatfacto` with `--downscale-factor 1`, so it does not need `images_2`, `images_4`, and `images_8`.
> 
> ################
> 
> Hey, Sorry for my careless. In first attempt that I said before:
> 
> ```
> ns-process-data images --data IMAGES \
>     --output-dir IMAGES_MULTICAM \
>     --sfm-tool hloc --feature-type disk --matcher-type disk+lightglue \
>     --skip-images-processing --no-same-dimensions --no-use_single_camera_mode
> 
> cp -r IMAGES/* IMAGES_MULTICAM/images
> ```
> 
> I found 8th images in `ns-view` with `viser`, but it has wrong orientation. image
> 
> Take a look in portrait image, but in `viser`, it was rotated automatically. image
> 
> The `transforms.json` also does not translate it to potrait. image

Probably the parameter `--no_same_dimensions` is doing the rotation.

**Comment by ichsan2895** [comment]
1. I think `--no-same-dimensions` is intended when we use different resolution images so the `transforms.json` creates resolution attributes for each image (instead of general attribute like colmap). But, I just know it possibly doing rotation too.

My recommendation @dberga : auto rotation has another flag like `--auto-rotation` & `--no-auto-rotation`. Then `--no-same-dimensions` for using different images resolution.

2. Run this command:
```
ns-process-data images --data IMAGES \
    --output-dir IMAGES_MULTICAM \
    --sfm-tool hloc --feature-type disk --matcher-type disk+lightglue \
    --skip-images-processing --same-dimensions --no-use_single_camera_mode

cp -r IMAGES/* IMAGES_MULTICAM/images
```
The images still rotated.

**Comment by ichsan2895** [comment]
Interesting finding:
1. Run this command:
```
colmap model_converter \
    --input_path IMAGES_MULTICAM/colmap/sparse/0 \
    --output_path IMAGES_MULTICAM/colmap/sparse/0 \
    --output_type TXT
```
2. Open `IMAGES_MULTICAM/colmap/sparse/0/cameras.txt`
3. All of the images has same resolution (despite in input, there are 1 potrait & 7 landscape images). So the autorotation caused by `hloc` I think.
`--no-same-dimension` and `--same-dimensions` does not have effect afterall since the cameras.txt has same resolution.
```
# Camera list with one line of data per camera:
#   CAMERA_ID, MODEL, WIDTH, HEIGHT, PARAMS[]
# Number of cameras: 8
8 OPENCV 4624 3472 3422.1921813355152 3420.2523863833289 2312 1736 0.072982410485304544 -0.11388968316811775 0.0014774746720753768 -0.0015804304081276507
7 OPENCV 4624 3472 3402.8378682136849 3397.5772595708336 2312 1736 0.082576038894206491 -0.15068003202440697 0.0013543233908796246 -0.00059204086450159483
6 OPENCV 4624 3472 3418.6384711506989 3412.8018492685578 2312 1736 0.055607240161820254 -0.061799060016399641 0.0010211725425041155 -0.00068701512432646462
5 OPENCV 4624 3472 3409.6767612744511 3419.3946673559321 2312 1736 0.07269686822641859 -0.091214819120700077 0.0021359424701642396 -0.002934762972419369
4 OPENCV 4624 3472 3398.1794370163429 3403.0184384187055 2312 1736 0.060456096642486239 -0.062672484962208674 0.00038078816854184716 -0.0023915384411823063
3 OPENCV 4624 3472 3373.4672113436777 3393.7038187398966 2312 1736 0.070667709654399016 -0.089770482952795674 0.00017934005012779037 -0.0030718905642279325
2 OPENCV 4624 3472 3390.3063822725585 3399.2824954832113 2312 1736 0.077084670017596793 -0.10620337151668688 0.0013100723624426642 -0.004853945126040737
1 OPENCV 4624 3472 3375.7035440085774 3104.9692601554093 2312 1736 -0.17221875899859262 -0.23014876258465428 -0.0038754464455599685 -0.0051693569426345893
```

**Comment by dberga** [comment]
The images in the output `IMAGES_MULTICAM` are rotated ? 

- (this is the preprocessing step) With `same_dimensions=True` should not rotate to the output instead it copies the images (https://github.com/nerfstudio-project/nerfstudio/blob/main/nerfstudio/process_data/process_data_utils.py#L276). Put a breakpoint there and check the output.
- in case the output images (IMAGES_MULTICAM) appear rotated in the output, in https://github.com/nerfstudio-project/nerfstudio/blob/main/nerfstudio/process_data/process_data_utils.py#L309 Rarely but could be https://trac.ffmpeg.org/ticket/4612 ? Put a breakpoint there in case it stops and check the output. 
- in case the output images (IMAGES_MULTICAM) do not appear rotated but the hloc/colmap data appears rotated, then it would be indeed an hloc bug???

**Pull request merged**